### ISSUE TITLE
Search word by word, instead of full phrase

### ISSUE THREAD

**Issue by ktaragorn** [select2/select2#1609]
I understand one can do this by a custom matcher, but it feels very necessary when you are searching among elements longer than a couple of words

For e.g

text - the quick brown fox
term - qui fo  (quick fox)

Currently - No match

with custom matcher they match.

the matcher is 

```
matcher: function(term, text) {
      var has = true;
      var words = term.toUpperCase().split(" ");
      for (var i =0; i < words.length; i++){
        var word = words[i];
        has = has && (text.toUpperCase().indexOf(word) >= 0); 
      }
      return has;
}
```

This shouldn't be a default for performance reasons (actually if it scales, perhaps it isn't too bad as a default), but it can be offered as a boolean flag to the constructor.

Basically what I am curious about is that is there any reason this isnt in already? The only thing I can think of is that selecting from multiword elements is a lot rarer than I am assuming.

**Comment by LiamMcArthur** [comment]
Thanks for that @ktaragorn - that matcher was just what I was looking for as I needed more of a "fulltext" search.

**Comment by ktaragorn** [comment]
@LiamMcArthur Glad I could help!

**Comment by LiamMcArthur** [comment]
This was working perfectly for me in 3.5.4, but since upgrading to 4.0.3 (as part of the WooCommerce version 3.0.0 update), it's not stopped working.

**Comment by akz92** [comment]
Apparently the matcher params changed, I adapted OP's code to work with select2's latest release:

```
$.fn.select2.defaults.set('matcher', function(params, data) {
  // If there are no search terms, return all of the data
  if ($.trim(params.term) === '') {
    return data;
  }

  // Do not display the item if there is no 'text' property
  if (typeof data.text === 'undefined') {
    return null;
  }

  var words = params.term.toUpperCase().split(" ");

  for (var i = 0; i < words.length; i++) {
    if (data.text.toUpperCase().indexOf(words[i]) < 0) {
      return null;
    }
  }

  return data;
});
```

PS: This changes the global select2 matcher.

**Comment by mbdesign** [comment]
I Try with latest release but not working.

i want to search as sample "New York City " like that  "Ne Ci" . but no results?

**Comment by C3maLi** [comment]
> En son sürümle çalışıyorum ama çalışmıyor.
> 
> "Ne Ci" gibi "New York City" örneğini aramak istiyorum. ama sonuç yok mu?

+++

**Comment by C3maLi** [comment]
> I Try with latest release but not working.
> 
> i want to search as sample "New York City " like that "Ne Ci" . but no results?

+1

**Comment by paarthpm** [comment]
> Apparently the matcher params changed, I adapted OP's code to work with select2's latest release:
> 
> ```
> $.fn.select2.defaults.set('matcher', function(params, data) {
>   // If there are no search terms, return all of the data
>   if ($.trim(params.term) === '') {
>     return data;
>   }
> 
>   // Do not display the item if there is no 'text' property
>   if (typeof data.text === 'undefined') {
>     return null;
>   }
> 
>   var words = params.term.toUpperCase().split(" ");
> 
>   for (var i = 0; i < words.length; i++) {
>     if (data.text.toUpperCase().indexOf(words[i]) < 0) {
>       return null;
>     }
>   }
> 
>   return data;
> });
> ```
> 
> PS: This changes the global select2 matcher.

This works. Thanks!!
